x/ccv/provider/keeper/validator_set_update.go

```diff
@@ -0,0 +1,218 @@
+func (k Keeper) DeleteAllEpochValidators(
+	ctx sdk.Context,
+	chainID string) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.OptedInBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+
+	var keysToDel [][]byte
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		keysToDel = append(keysToDel, iterator.Key())
+	}
+	for _, delKey := range keysToDel {
+		store.Delete(delKey)
+	}
+}
+
+// IsEpochValidator returns `true` if the validator with `providerAddr` is set in and `false` otherwise
+func (k Keeper) IsEpochValidator(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.OptedInKey(chainID, providerAddr.ToSdkConsAddr())) != nil
+}
+
+// GetAllEpochValidators returns all the epoch validators on chain `chainID`
+func (k Keeper) GetAllEpochValidators(
+	ctx sdk.Context,
+	chainID string) (optedInValidators []types.EpochValidator) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.OptedInBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		iterator.Value()
+		var optedInValidator types.EpochValidator
+		if err := optedInValidator.Unmarshal(iterator.Value()); err != nil {
+			panic(fmt.Errorf("failed to unmarshal CurrentEpochOptedInValidator: %w", err))
+		}
+		optedInValidators = append(optedInValidators, optedInValidator)
+	}
+
+	return optedInValidators
+}
+
+// ComputeNextEpochValidators returns the next validator set that is
+// responsible for validating on a consumer chain.
+func (k Keeper) ComputeNextEpochValidators(
+	ctx sdk.Context,
+	chainID string,
+	currentValidators []types.EpochValidator,
+	stakingValidators []types2.Validator,
+) []types.EpochValidator {
+	isCurrentValidator := make(map[string]types.EpochValidator)
+	for _, val := range currentValidators {
+		// `currentPublicKey` is the currently used key by validator `val` when validating on the consumer chain
+		isCurrentValidator[string(val.ProviderConsAddr)] = val
+	}
+
+	var nextValidators []types.EpochValidator
+	for _, val := range stakingValidators {
+		// get next voting power and the next consumer public key
+		nextPower := k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
+		consAddr, err := val.GetConsAddr()
+		if err != nil {
+			// this should never happen but is recoverable if we exclude this validator from the `nextValidators`
+			k.Logger(ctx).Error("could not get consensus address of validator (%+v): %w", val, err)
+			continue
+		}
+		nextConsumerPublicKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, types.NewProviderConsAddress(consAddr))
+		if !found {
+			k.Logger(ctx).Error("could not retrieve public key for validator (%+v)", val)
+			// if no consumer key assigned then use the validator's key itself
+			nextConsumerPublicKey, err = val.TmConsPublicKey()
```

Comment:
Yes. This is actually tested in `TestComputeNextEpochValidators` and if we were to remove this assignment, the test would fail. This happens because the test also checks against a validator that hasn't set a consumer key.